src/conn/conn_compact.c

```diff
@@ -89,45 +144,19 @@ __compact_server(void *arg)
         if (!running)
             continue;
 
-        /* Open a metadata cursor. */
-        WT_ERR(__wt_metadata_cursor(session, &cursor));
-
-        cursor->set_key(cursor, uri);
-        WT_ERR(cursor->search_near(cursor, &exact));
-
-        /* Make sure not to go backwards. */
-        if (exact <= 0)
-            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
-
-        /* Find a table to compact. */
-        while (ret == 0) {
-            WT_ERR(cursor->get_key(cursor, &key));
-            /* Check we are still dealing with keys which have the right prefix. */
-            if (WT_PREFIX_MATCH(key, prefix)) {
-                /* There are files that should not be compacted. */
-                if (!WT_STREQ(key, WT_HS_URI))
-                    /* FIXME-WT-11343: check if the table is supposed to be compacted. */
-                    break;
-            } else {
-                ret = WT_NOTFOUND;
-                break;
-            }
-            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
-        }
+        /* Find the next URI to compact. */
+        __wt_free(session, next_uri);
+        WT_ERR_NOTFOUND_OK(__find_next_uri(session, uri, &next_uri), true);
 
         /* All the keys with the specified prefix have been parsed. */
         if (ret == WT_NOTFOUND) {
-            WT_ERR(__wt_metadata_cursor_release(session, &cursor));
             full_iteration = true;
             continue;
         }
 
-        /* Make a copy of the key as it can be freed once the cursor is released. */
+        /* Use the retrieved URI. */
         __wt_free(session, uri);
-        WT_ERR(__wt_strndup(session, key, strlen(key), &uri));
-
-        /* Always close the metadata cursor. */
-        WT_ERR(__wt_metadata_cursor_release(session, &cursor));
+        WT_ERR(__wt_strndup(session, next_uri, strlen(next_uri), &uri));
```

Comment:
Do you need a new copy of the string, or can you just say `uri = new_uri` here?

Reply:
I have the feeling I can do better than the current code and head towards what you are suggestion, I will double check.

Reply:
I have used scratch buffers in b844b8bd17bba1da62201fa42bd83d8c4eb8a76f and b9ac309088a52785b995cd3a568dc6e9ce9f5a9e. What do you think? This avoids the calls to `free` and `strndup` in the loop. The buffer grows as needed and we only need to allocate/free them outside the loop.

Reply:
Added a fix regarding the string length: 9abaf720b4bfff985068fb5a532c1bbda5056a58.